src/mumble/dialogs/PositionalAudioViewer.cpp

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#include "PositionalAudioViewer.h"
+#include "ui_PositionalAudioViewer.h"
+
+#include "PluginManager.h"
+#include "Global.h"
+
+PositionalAudioViewer::PositionalAudioViewer(QWidget *parent) : QDialog(parent), m_ui(new Ui::PositionalAudioViewer) {
+	m_ui->setupUi(this);
+
+	connect(&m_timer, &QTimer::timeout, this, &PositionalAudioViewer::update);
+
+	m_timer.start(100);
+}
+
+PositionalAudioViewer::~PositionalAudioViewer() {
+}
+
+void PositionalAudioViewer::update() {
+	auto pluginManager = Global::get().pluginManager;
+	if (!pluginManager) {
+		return;
+	}
+
+	pluginManager->fetchPositionalData();
+
+	const auto &data = pluginManager->getPositionalData();
+
+	updatePlayer(data);
+	updateCamera(data);
+
+	m_ui->context->setPlainText(data.getContext());
+	m_ui->identity->setPlainText(data.getPlayerIdentity());
+}
+
+void PositionalAudioViewer::updatePlayer(const PositionalData &data) {
+	const auto pos  = data.getPlayerPos();
+	const auto dir  = data.getPlayerDir();
+	const auto axis = data.getPlayerAxis();
```

Comment:
`auto` hides variable type

Plus copying this data seems unnecessary. A reference should be just fine :thinking:

Reply:
Is the data actually copied?

Reply:
Yes, because you don't use `auto &`. I am somewhat confident that `auto` does not resolve to references automatically :thinking:

Reply:
It does when the function returns a reference.

In this case the `get` functions return an object that is allocated on the stack. As a result, the variable is copied on assignment.

Specifying `&` actually appears to be less efficient: https://godbolt.org/z/TPcEfxsef

Clang 12.0.0:
```asm
data():                               # @data()
        push    rbp
        mov     rbp, rsp
        mov     rax, rdi
        pop     rbp
        ret
non_ref():                            # @non_ref()
        push    rbp
        mov     rbp, rsp
        sub     rsp, 48
        lea     rdi, [rbp - 40]
        call    data()
        add     rsp, 48
        pop     rbp
        ret
ref():                                # @ref()
        push    rbp
        mov     rbp, rsp
        sub     rsp, 48
        lea     rdi, [rbp - 48]
        call    data()
        lea     rax, [rbp - 48]
        mov     qword ptr [rbp - 8], rax
        add     rsp, 48
        pop     rbp
        ret
```

GCC 11.1:
```asm
data():
        push    rbp
        mov     rbp, rsp
        mov     QWORD PTR [rbp-8], rdi
        nop
        mov     rax, QWORD PTR [rbp-8]
        pop     rbp
        ret
non_ref():
        push    rbp
        mov     rbp, rsp
        sub     rsp, 48
        lea     rax, [rbp-48]
        mov     rdi, rax
        call    data()
        nop
        leave
        ret
ref():
        push    rbp
        mov     rbp, rsp
        sub     rsp, 48
        lea     rax, [rbp-48]
        mov     rdi, rax
        call    data()
        lea     rax, [rbp-48]
        mov     QWORD PTR [rbp-8], rax
        nop
        leave
        ret
```